### PR TITLE
Add sympathy dev warning

### PR DESCRIPTION
**Pull request by ddc22**
#### Proposed Changes
Adds a sympathy dev warning so developers are warned of the random local storage wipe out.

### Attempt 2 - Adding to Initial redux state of notices
image

### Attempt 1 - Direct DOM Attachment and removal 7fdb1c2263
image

#### Testing Instructions
- Keep refreshing a few times (At least 4 times) window and make sure the above notice appears. (Try  logged in and logged out(signup flow))

#### Pre-merge Checklist

- [ ] [Have you written new tests](https://wpcalypso.wordpress.com/devdocs/docs/testing/index.md) for your changes?
- [ ] Have you tested the feature in Simple (P9HQHe-k8-p2), Atomic (P9HQHe-jW-p2), and self-hosted Jetpack sites (PCYsg-g6b-p2)?
- [ ] Have you checked for TypeScript, React or other console errors?
- [ ] Have you used memoizing on expensive computations? More info in [Memoizing with create-selector](https://github.com/Automattic/wp-calypso/blob/trunk/packages/state-utils/src/create-selector/README.md) and [Using memoizing selectors](https://react-redux.js.org/api/hooks#using-memoizing-selectors) and [Our Approach to Data](https://github.com/Automattic/wp-calypso/blob/trunk/docs/our-approach-to-data.md)
- [ ] Have we added the "[Status] String Freeze" label as soon as any new strings were ready for translation (p4TIVU-5Jq-p2)?

Related to #

### CODE REVIEW

**Comment by dmsnell** [comment]
I love the idea of adding a UI element when the persisted cache is cleared, but I'm curious if direct DOM manipulation is the best way. Did you look into using the ENV badge, suggested in the original PR (#14121)?

Alternatively, I bet we could set some app state and render the warning in React with existing notices.

**Comment by ddc22** [comment]
If the ENV is rendered from the backend as mentioned by you I don't think it makes sense to make any changes there for a frontend state clearing dev feature.

I was reluctant to use Redux because we are already inside redux internals and it could cause weird side effects if we try to dispatch while getting the 'initial state'. I will try and see however if it is possible to do something with state.
@dmsnell

**Comment by ddc22** [comment]
@dmsnell  So I did manage to add this to our state but I getting the following error. Obviously this is because I am doing something underhanded and our global react-redux does not contain the notices reducer plugged into it initially.

image


@sgomes I know that you have left a warning to not at our global reducer eagerly, But I feel like notices is something truly global and and can be added here. Is it ok if I add the notices reducer to the global state?

https://github.com/Automattic/wp-calypso/blob/7fdb1c226376d785d9da868b24be9497f93b248a/client/state/reducer.js#L13-L22

P.S. This seems to be breaking our tests which might be a red flag for this implementation. But I would like to know what you think about the approach before we proceed any further.
CC - @tyxla @sgomes @dmsnell

File: client/state/initial-state.js
Comment (dmsnell) [inline client/state/initial-state.js]: was this change intentional? I wonder if this is the reason for the console warning you are seeing.

File: client/state/initial-state.js
Comment (dmsnell) [inline client/state/initial-state.js]: if I understand this properly we're _always_ setting the notification here even when persisted state _wasn't_ cleared, unless someone intentionally disables the sympathy setting.

I was originally thinking more of setting some value in `initialState` if we happened to wipe out the persisted state, and then after app boot display the notice. this would both be easier to test and also should trigger the notification to appear rather than to be evidently there from the beginning (losing the flash that draws attention).

that is, we can dispatch the notification from normal code somewhere in a normal React component added in DEV environments.

the same would work if we modify the server code to send the signal to clear persistence. if we want the ENV badge to display that then I guess we don't need to interact with Redux, or maybe we have some examples of settings already being set for the ENV badge and can follow their example.

it's okay to modify the server and have it perform the random test: "should the client wipe out persistence on this page load?" and then have the client respond to it.

Comment (tyxla) [reply client/state/initial-state.js]: I agree with Dennis here. From a devex standpoint, it makes more sense to display the notice later, after the page has fully loaded, so it will draw attention. 

One way to achieve that could be on the dev environment to set a flag in the app context and later read that flag, and if it's set, show the notice.

File: client/state/initial-state.js
Comment (ddc22) [reply client/state/initial-state.js]: This is a mistake.

File: client/state/initial-state.js
Comment (ddc22) [reply client/state/initial-state.js]: Hi @dmsnell I moved the state to a flag and dispatched a notice when the layout was mounted. Can you please have another look.

File: client/layout/logged-out.jsx
Comment (ddc22) [inline client/layout/logged-out.jsx]: Added a delay so the notice appears after the UI is mostly loaded.

File: client/layout/logged-out.jsx
Comment (ddc22) [inline client/layout/logged-out.jsx]: Change was added to both default and logged out layouts instead of adding into SSR or normal loaded app versions.

File: client/state/initial-state.js
Comment (dmsnell) [reply client/state/initial-state.js]: @jdc91 the issue isn't how the value is stored. if I'm understanding this properly we're still relying on the `config.isEnabled( 'force-sympathy' )` flag but that flag isn't sufficient to indicate when state is cleared; in fact, most of the time that state is cleared, that flag is not enabled.

right now `shouldAddSympathy()` returns whether or not state should be cleared, and it does so randomly, meaning that we can't call it multiple times and know. (we haven't needed to call it multiple times because it feeds into the initial app state).


I also don't believe that we need all the Redux actions for this, as we have an opportunity right here in this function to return an indication that we cleared. something like this…

```ts
if ( shouldAddSympathy() ) {
	console.log( … );
	return { startupFlags: { clearedInitialPersistedState: true } };
}
```

and now we know this value will never change and we can read it from state directory, or use a selector. point is that we can get it into state without a lot of overhead because it's only set once and that at boot up.

if we can start indicating from server code whether or not to clear then all we need to do is modify `shouldAddSympathy()` to take that into account. for now though I believe part of this PR could be removed without taking away any of the desired behavior.

File: client/layout/index.jsx
Comment (tyxla) [inline client/layout/index.jsx]: Is there any chance we can abstract that to a separate component that we conditionally load only on the environments where sympathy mode is enabled? The layout component is quite huge already so anything that can be abstracted away from it is a small win.

File: client/layout/logged-out.jsx
Comment (tyxla) [inline client/layout/logged-out.jsx]: Same as above in terms of abstracting away from the component. Even if it is not async loaded (since logged-out is used in SSR), from the perspective of code organization, readability and maintainability standpoint.

File: client/state/action-types.ts
Comment (tyxla) [inline client/state/action-types.ts]: Nit: the spaces should likely be removed to keep consistency.

File: client/state/startup-flags/actions.ts
Comment (tyxla) [inline client/state/startup-flags/actions.ts]: Using Redux for this seems a bit over-engineered to me. Wouldn't it be simpler if we use a flag inside `persisted-state` and updating that flag when persisted state is cleared? Then using a simple selector-like function that returns the flag?

File: client/state/reducer.js
Comment (tyxla) [inline client/state/reducer.js]: Ideally, we want to reduce this list to zero, not grow it further 😉

File: client/state/initial-state.js
Comment (tyxla) [inline client/state/initial-state.js]: It seems like `isStateRandomlyCleared` mirrors whether the sympathy mode is enabled, and not whether persisted state was cleared as a consequence of sympathy mode running.

File: client/layout/index.jsx
Comment (tyxla) [inline client/layout/index.jsx]: Not sure this makes sense, but do we need a flag that we displayed this notice already? Just in case we mount this component again in the same session.

File: client/layout/logged-out.jsx
Comment (tyxla) [reply client/layout/logged-out.jsx]: Are we certain that this delay will be enough for smaller machines? Should we use different means to ensure that UI has been loaded?

Comment (ddc22) [reply client/layout/logged-out.jsx]: This is not an end user facing feature and I didn't think it's vital to ensure this. Maybe I can increase the duration the notice lingers on before auto disappearing?

File: client/layout/index.jsx
Comment (ddc22) [reply client/layout/index.jsx]: Good idea I will give this a go.

File: client/state/startup-flags/actions.ts
Comment (ddc22) [reply client/state/startup-flags/actions.ts]: I didn't want to use the browser store because we don't need the flag to be persistent. And I was modelling this state slice as a place to maintain any future such flags. But if you think it's better to just move it to persistent storage I can go ahead with it.

File: client/state/reducer.js
Comment (ddc22) [reply client/state/reducer.js]: Oh I thought this slice sort of makes sense to live in this file because it is meant as a global startup slice. Lazy registering using `registerReducer` felt unnecessary since this slice will anyways be loaded on every boot by definition. WDYT?

File: client/state/initial-state.js
Comment (ddc22) [reply client/state/initial-state.js]: - Actually it does mean that. If you look a few lines above the state was indeed cleared.
https://github.com/Automattic/wp-calypso/blob/bb7777c98f1ee12d4ddb40009d89081a1ab4d18d/client/state/initial-state.js#L178-L182

- If the `force-sympathy` flag wasn't enabled (which means the state will always be cleared), And the state was actually cleared. It means that the random clearing logic is in effect. The actual line of code lives below.
https://github.com/Automattic/wp-calypso/blob/bb7777c98f1ee12d4ddb40009d89081a1ab4d18d/client/state/initial-state.js#L59-L64

Or am I missing something? If not should I probably add a comment here explaining this?

File: client/layout/index.jsx
Comment (ddc22) [reply client/layout/index.jsx]: As soon as the notice is displayed we update this Redux state `isStateRandomlyCleared` to be false. 

https://github.com/Automattic/wp-calypso/blob/274d355fc8e667a443907956d2f86f96e0d06477/client/state/startup-flags/reducer.ts#L11-L13

So unless the whole app boots up again this state will remain false even if this component remounts.

P.S. The reducer implementation was wrong and I fixed it in 274d355fc8e667a443907956d2f86f96e0d06477

File: client/state/reducer.js
Comment (sgomes) [reply client/state/reducer.js]: Hey! Sorry, just came across this now.

Must it really be loaded on every boot? I was under the impression this was only needed in development builds, or rather, in environments where sympathy is enabled.

File: client/state/initial-state.js
Comment (dmsnell) [reply client/state/initial-state.js]: yeah I think I missed that this was already inside the `if ( shouldAddSympathy() )` block. in that case, why are we checking the flag at all instead of directly returning `{ persistedStateWasCleared: true }`?

Comment (tyxla) [reply client/state/initial-state.js]: I understand what you mean 👍 It makes sense to not show the notice if `force-sympathy` is specifically enabled.

File: client/state/reducer.js
Comment (tyxla) [reply client/state/reducer.js]: I really think we can do it in a different way and not use Redux at all for this. I mentioned using a flag in the `persisted-state` module already, I'll leave a comment there to clarify what I mean.

File: client/state/startup-flags/actions.ts
Comment (tyxla) [reply client/state/startup-flags/actions.ts]: I'm not saying to persist to the browser; but just to have a simple flag in the `persisted-state` module. That flag would indicate whether the user was notified, and would default fo `false`, then we'd set it to `true` inside the component when we dispatch the notice. The behavior is simple enough and we want to do it just once, so that flag should do the job. I'm really hesitant to use Redux and specifically add a new key that loads for everyone for such a simple "toggle-like" behavior.

Comment (ddc22) [reply client/state/startup-flags/actions.ts]: Got it @tyxla let me give it a try!

File: client/state/initial-state.js
Comment (ddc22) [inline client/state/initial-state.js]: @tyxla  I used `browser-storage` here instead of `persisted-state` because that seemed to be tailored/targeted for redux and query data.
https://github.com/Automattic/wp-calypso/blob/e7af7f07ed4a6466d2be0b5e5ccd31dc1b6d4aee/client/state/persisted-state.js#L17

So I directly used the [browser-state](https://github.com/Automattic/wp-calypso/tree/v16.3.0-beta1/client/lib/browser-storage) abstraction.

File: client/layout/index.jsx
Comment (ddc22) [reply client/layout/index.jsx]: Same is done when using browser-storage.
https://github.com/Automattic/wp-calypso/blob/e7af7f07ed4a6466d2be0b5e5ccd31dc1b6d4aee/client/components/sympathy-dev-warning/index.tsx#L30

File: client/state/reducer.js
Comment (ddc22) [reply client/state/reducer.js]: This has been removed.

File: client/state/startup-flags/actions.ts
Comment (ddc22) [reply client/state/startup-flags/actions.ts]: I have removed any redux involvement, but I used browser-storage instead of persisted-state as explained [here](https://github.com/Automattic/wp-calypso/pull/69565#discussion_r1016817750).

**Comment by ddc22** [comment]
@tyxla I have done the changes as you requested. Can you have another look :)

**Review by dmsnell — COMMENTED** [review]
Well I thought I left a comment before but apparently never hit submit.

The thing to be aware of with `config.isEnabled( 'force-sympathy' )` is that it should only ever be set if someone intentionally loaded the site with an ENV set or with a query arg. I would imagine it's hard to not realize that's there if it takes that active choice to force it.

The random bit is the part that's suprising.

File: client/state/initial-state.js
Comment (ddc22) [reply client/state/initial-state.js]: > I also don't believe that we need all the Redux actions for this, as we have an opportunity right here in this function to return an indication that we cleared. something like this…

I did follow that exact approach of returning the state as shown below in the previous commit.
https://github.com/Automattic/wp-calypso/blob/13740395f29f1ac74e771cd6f73b80f072b0a7a3/client/state/initial-state.js#L181-L183

I only added 1 action to **disable** this state once the notice was published. This was mainly so that any remount of the layout will not trigger a re-appearance of that warning.

https://github.com/Automattic/wp-calypso/blob/13740395f29f1ac74e771cd6f73b80f072b0a7a3/client/state/startup-flags/actions.ts#L3-L8

The other action was merely a bit of DRY for the notices calling action so that it can be reused easily in both logged in and logged out contexts.
https://github.com/Automattic/wp-calypso/blob/13740395f29f1ac74e771cd6f73b80f072b0a7a3/client/state/startup-flags/actions.ts#L10-L20

---
> @dmsnell @jdc91 the issue isn't how the value is stored. 

Actually @tyxla raised some valid concerns about this [here](https://github.com/Automattic/wp-calypso/pull/69565#discussion_r1015913177). Therefore this was moved into local storage completely.

File: client/state/initial-state.js
Comment (ddc22) [reply client/state/initial-state.js]: @dmsnell 
> in that case, why are we checking the flag at all instead of directly returning { persistedStateWasCleared: true }

A notice is already bit of a hindrance to DX. If a developer is deliberately enabling sympathy, we do not need to state the obvious by showing them a warning because they know what they are doing.

The main point of having a warning like this was to provide some feedback IFF a state was cleared **randomly**. And this is the exact case I am trying to capture/isolate by checking the flag. And then and only then will we be providing a warning.

https://github.com/Automattic/wp-calypso/blob/e7af7f07ed4a6466d2be0b5e5ccd31dc1b6d4aee/client/state/initial-state.js#L184

In other words if the state was cleared randomly
1 - [shouldAddSympathy returns true](https://github.com/Automattic/wp-calypso/blob/e7af7f07ed4a6466d2be0b5e5ccd31dc1b6d4aee/client/state/initial-state.js#L174)
2 - [The force-sympathy feature flag should **not** have been set](https://github.com/Automattic/wp-calypso/blob/e7af7f07ed4a6466d2be0b5e5ccd31dc1b6d4aee/client/state/initial-state.js#L184)

File: client/state/initial-state.js
Comment (tyxla) [reply client/state/initial-state.js]: Oh, I didn't mean to persist that piece of data, but rather to have a simple variable as a flag that just lives inside that `persisted-state` module, because that way we can set that flag when we clear the state since the `clearPersistedState()` function lives there. What do you think about that?

Comment (ddc22) [reply client/state/initial-state.js]: I feel like that approach is a bit hacky than the current approach. And it would be confusing to see a flag like that inside a module meant to handle persisted-state data. 

Do you think the current approach doesn't make sense? If crowding broswer storage is a problem I can actually remove the state once we inform the user. It would be like nothing happened.

Comment (tyxla) [reply client/state/initial-state.js]: Could you elaborate on what makes it hacky from your perspective? Also, it's related to persisted state module itself, so I see a good reason to have it in that module, so maybe you could elaborate on that as well? I'm just seeking the simplest possible way to do this, and I don't see a simpler one than the flag one.

Comment (ddc22) [reply client/state/initial-state.js]: As I mentioned above, it would be confusing to see a flag like that inside a module meant to handle persisted-state data. The persisted state module does not need to know about a random clearing devx feature. The concept of sympathy is quite a unique use case with a backstory and valid motivation behind it.However this feature while essential takes some time to wrap your head around.

Now let's say a developer wants to do some changes to the `persisted-state` module. We have now exposed them to a one variable related to the sympathy concept. Unnecessary unrelated complexity is exposed where it is not needed.

Nevertheless, why do you think the current approach is more complex? From a simple LOC standpoint the solution you propose would be equal to mine or sometimes be higher. And from a cohesion stand point the solution I propose will keep like things together. The code related to informing the developer about sympathy will only live in.
1 - Where sympathy is actually happening
2 - In the component where we publish the sympathy warning

Comment (ddc22) [reply client/state/initial-state.js]: I think this is also more easier to code-read, from a traceability standpoint, for someone trying to understand the symphony dev notice feature. You just have to do a search for the storage key and you can instantly find the exact two places this state is accessed and updated to understand what is happening.

If we go with the other approach the user will have to click across multiple times with getters and setters to figure out what is happening.

Comment (tyxla) [reply client/state/initial-state.js]: FWIW the "where it lives" part is not so relevant for the discussion - we could move the flag to where the rest of the sympathy mode code lives and what I suggested would still stand, and I do feel what I suggest would be simpler, mostly because we don't need to clear the flag after we've displayed a notice (which is something necessary for your approach).

I don't have so strong feelings about it, but I do feel that using browser storage is still more complex than it should be for such a simple thing. The main reason is that we don't really leverage the browser storage in the way we usually do - as a persistence layer, but rather we use it as a runtime store, and I don't see the need for that. I might be missing something, though - do let me know if that's the case!

In any case, the current state of your proposal looks much better than the initial one, and I don't have many more thoughts to add.

File: client/layout/index.jsx
Comment (tyxla) [inline client/layout/index.jsx]: One thing I'd still like to do is conditionally render this only if this is the development environment. Feel free to abstract this into the component itself, or to put it in the layout components, but I believe that's one of the last pieces we're missing here.

Comment (ddc22) [reply client/layout/index.jsx]: Good call! I have done that here bf7d822be9545857711b1737bb09c8655eb056c6

File: client/state/initial-state.js
Comment (dmsnell) [reply client/state/initial-state.js]: @jdc91 I should have written "my issue isn't how the value is stored" or "the issue I mentioned above isn't how it's stored"

since you've address what I was asking about there's not need to reply; but I wanted to clear up my comment.

File: client/state/initial-state.js
Comment (dmsnell) [reply client/state/initial-state.js]: @jdc91 great, and I apologize if I'm overlooking things I should be seeing (I _am_ a bit distracted at a meetup right now).

I may actually be finding the naming confusing here and the use of the config value. it may still be that we can rearrange this in a clearer way.

```ts
if ( shouldAddSympathy() ) {
	…

	…

	if ( config.isEnable( 'force-sympathy' ) ) {
		// If we're forcing the state-clearing we don't
		// have to announce it, because someone intentionally
		// forced it and should know why it's cleared.
		return;
	}

	return { randomlyClearedState: true };
}
```

```ts
if ( shouldAddSympathy() ) {
	…

	// only alert if the clearing wasn't intentionally set by a dev
	const wasRandomlyCleared = ! config.isEnabled( 'force-sympathy' );

	return { wasRandomlyCleared… };
}
```

```ts
if ( shouldAddSympathy() ) {
	…

	return {
		didClearPersistedState: true,
		didRandomlyClearPersistedState: ! config.isEnabled( 'force-sympathy' )
	}
}
```

in these three forms I'm trying to explore a way to express what you are describing in words that doesn't couple the implementation logic of the alert in with the particular about the config flag. at least here there's only an indirect connection with the way the code is written and the intention to flag the announcement.

it seems fine to run any of these things and at this point I don't want to block the PR because of it, but I think we can remove some confusion simply by giving new names to this idea (or these two ideas: was cleared? was cleared by a random process?)

**Comment by southp** [comment]
Thanks for the amazing collaboration on this, @jdc91 , @dmsnell , @tyxla . I've tested it and it works beautifully for me. The code looks great after reading through the discussion that led us to this point, but it's so deep that I'm not confident that I can approve. 

@dmsnell @tyxla when you have some time, I'd be grateful if either of you can help us make the final call here 🙇🏼

**Review by tyxla — COMMENTED** [review]
I have nothing else to add, but I'll leave @dmsnell to make the final approval here! Thanks for working on it @jdc91!

**Review by dmsnell — APPROVED** [review]
sorry everyone; I wrote comments again without hitting "Summit Review"

I'm going to optimistically approve this, but I would ask that you pay attention to my latest comments (which I thought you saw yesterday but realize now you didn't). In summary: I have no opinion on the component which shows the message, but I find the naming around `randomlyClears` confusing and think that could be cleared up simply by calling it `didRandomlyClear` (or something like that) _or_ by tracking if it cleared at all and then secondarily, if it cleared randomly or by force.

Awesome work, and awesome work enduring this lengthy review 😆 

You are making things better for developers and for the customers.

**Comment by ddc22** [comment]
Hi @dmsnell thank's for your feedback on making this more readable. I went a bit overboard 😅  but I think it is warranted given that there is some complexity here.
https://github.com/Automattic/wp-calypso/blob/48d430d7d2acbca25373919f76f6c5ea3c4cbc78/client/state/initial-state.js#L181-L201

**Comment by dmsnell** [comment]
> I went a bit overboard

no it's great! documentation is there _for what isn't there_ and I think the empty if clause is fine.

longest review for the smallest change 😄 

🙌

**Pull request merged**
